File: app/schemas/compare.py
# app/schemas/compare.py
#
# Pydantic schemas for clause comparison feature (Day 8).
# Compares clauses across two documents for similarity.

from pydantic import BaseModel, Field
from typing import List, Optional


class CompareRequest(BaseModel):
    """
    Request schema for comparing two documents.
    """

    document_id_1: int = Field(
        ..., description="First document ID to compare"
    )
    document_id_2: int = Field(
        ..., description="Second document ID to compare"
    )
    threshold: float = Field(
        default=0.7,
        ge=0.0,
        le=1.0,
        description="Similarity threshold (0-1, default 0.7)"
    )


class ClauseComparison(BaseModel):
    """
    Comparison result for a single clause pair.
    """

    clause_1: str = Field(description="First clause text")
    clause_2: str = Field(description="Second clause text")
    similarity_score: float = Field(description="Similarity score (0-1)")
    document_1_chunk_id: Optional[int] = Field(
        None, description="Chunk ID from document 1"
    )
    document_2_chunk_id: Optional[int] = Field(
        None, description="Chunk ID from document 2"
    )
    document_1_page: Optional[int] = Field(
        None, description="Page number from document 1"
    )
    document_2_page: Optional[int] = Field(
        None, description="Page number from document 2"
    )


class CompareResponse(BaseModel):
    """
    Response schema for comparison results.
    """

    document_1_id: int = Field(description="First document ID")
    document_2_id: int = Field(description="Second document ID")
    document_1_filename: str = Field(description="First document filename")
    document_2_filename: str = Field(description="Second document filename")
    total_comparisons: int = Field(description="Total clause pairs compared")
    similar_clauses: List[ClauseComparison] = Field(
        description="List of similar clause pairs above threshold"
    )
    comparison_time_ms: float = Field(description="Comparison duration in ms")

File: frontend/compare.js
// frontend/compare.js
// Compare logic - can be called from chat view when agent triggers comparison

import { API } from './api.js';

async function compareDocuments(docId1, docId2, threshold = 0.7) {
    try {
        const response = await API.compareDocuments(docId1, docId2, threshold);

        if (response.ok) {
            return await response.json();
        } else {
            const error = await response.json();
            throw new Error(error.detail || 'Comparison failed');
        }
    } catch (err) {
        throw new Error(err.message || 'Network error during comparison');
    }
}

function formatCompareResult(result) {
    if (!result || !result.similar_clauses) {
        return 'No comparison results available.';
    }

    const { similar_clauses, total_comparisons, document_1_filename, document_2_filename, comparison_time_ms } = result;

    let output = `**Comparison: ${document_1_filename} vs ${document_2_filename}**\n`;
    output += `Total comparisons: ${total_comparisons} | Time: ${comparison_time_ms}ms\n\n`;

    if (similar_clauses.length === 0) {
        output += 'No similar clauses found above the threshold.';
        return output;
    }

    output += `**Found ${similar_clauses.length} similar clause(s):**\n\n`;

    similar_clauses.forEach((clause, i) => {
        output += `${i + 1}. **Similarity: ${(clause.similarity_score * 100).toFixed(1)}%**\n`;
        output += `   Doc 1 (p.${clause.document_1_page}): "${clause.clause_1.substring(0, 100)}..."\n`;
        output += `   Doc 2 (p.${clause.document_2_page}): "${clause.clause_2.substring(0, 100)}..."\n\n`;
    });

    return output;
}

export { compareDocuments, formatCompareResult };